### ISSUE TITLE
AKS container insights shows secrets in cleartext

### ISSUE THREAD

**Issue by NileshGule** [Azure/AKS#2387]
**What happened**:
Deployed MSSQL 2019 container on ASK cluster with SA password set as environment variable from a secret
**What you expected to happen**:
Secrets should not be shown in cleartext. The container insights section shows the SA password in cleartext under environment variables. This defeats the whole purpose of storing credentials as secrets. I have observed the same behaviour at my workplace where we inject the secrets using Hashicorp Vault backed resource. Those are also shown in clear text. This is clearly a security violation.
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
The repo used for deploying the SQL Server 2019 container can be found on GitHub 
https://github.com/NileshGule/spring-boot-conference-app/blob/mssql-server/k8s/conferencedb/db-statefulset.yml

![image](https://user-images.githubusercontent.com/1298374/121793369-f83a5100-cc30-11eb-89b1-4168e3b93273.png)

**Comment by SatyKrish** [comment]
This is due to upstream kubernetes implementation of injecting secrets as environment variables. You can disable environmental variables collection for pods. Refer this:[how-to-disable-environment-variable-collection-on-a-container ](https://docs.microsoft.com/en-us/azure/azure-monitor/containers/container-insights-manage-agent#how-to-disable-environment-variable-collection-on-a-container)

**Comment by NileshGule** [comment]
@SatyKrish thanks for the quick response. It is ok to display the list of environment variables.
However, showing the secret in cleartext is not acceptable.

Here is an example from Octant which also shows the list of environment variables. But it does not show the values in cleartext.
![image](https://user-images.githubusercontent.com/1298374/121801506-3142e780-cc6a-11eb-9d0c-2d2a96940009.png)
 

If you  navigate to the link sqlsecret:sapassword which takes you to the Kubernetes secret details, the password is still protected
![image](https://user-images.githubusercontent.com/1298374/121801524-559ec400-cc6a-11eb-8020-fa21271f9b4a.png)

**Comment by justindavies** [comment]
Hi there, as @SatyKrish said standard Kubernetes secrets are base64 encoded, and are exported as environment variables which is why you are able to see the password when looking at the pod environment.  Whilst this may seem alarming, being able to view these is based on namespace permissions.  You can take a look at aad-pod-identity and the Key Vault CSI driver for alternatives that allow you to authenticate with Azure services or externalise sensitive information respectively.

**Comment by sindrel** [comment]
> Hi there, as @SatyKrish said standard Kubernetes secrets are base64 encoded, and are exported as environment variables which is why you are able to see the password when looking at the pod environment. Whilst this may seem alarming, being able to view these is based on namespace permissions. You can take a look at aad-pod-identity and the Key Vault CSI driver for alternatives that allow you to authenticate with Azure services or externalise sensitive information respectively.

Hi @justindavies,

By namespace permissions, are you referring to Kubernetes RBAC namespace permissions? I set up a test scenario where I stripped a user of any permissions in Kubernetes RBAC, but the user is still able to view metrics and environment variables (including secrets) on all containers in the cluster in the Insights view. I expected it to not be able to view any metrics or environment variables in the Insights view. Other Kubernetes Service views in the Portal seem to respect Kubernetes RBAC rules. 

From what I can tell, the Insights tool does not use the access rights of the logged-in user.

**Permissions set:**
Kubernetes RBAC: None
Azure IAM: Azure Kubernetes Service Cluster User Role (built-in)

The Insights view show metrics and environment variables for all containers in the cluster:
![image](https://user-images.githubusercontent.com/35308882/127131895-3b832954-bd2a-471f-bd8b-95382a772f6e.png)

Other Kubernetes Service views in the Portal seem to respect Kubernetes RBAC rules:
![image](https://user-images.githubusercontent.com/35308882/127131856-09d547e8-3950-4ca0-969b-00d410c1b408.png)

For reference, this is the output when attempting to list namespaces or other resources using `kubectl`:

```
$ kubectl get ns
Error from server (Forbidden): namespaces is forbidden: User "<redacted>@<redacted>.com" cannot list resource "namespaces" in API group "" at the cluster scope
```
```
$ kubectl get pods -n test-app
Error from server (Forbidden): pods is forbidden: User "<redacted>@<redacted>.com" cannot list resource "pods" in API group "" in the namespace "test-app"
```

The pods in `test-app` used here is listed in the Insights view, including env vars.

The cluster uses the native Azure AD integration with Kubernetes RBAC privileges mapped to AD security groups, as per Azure documentation.
